proxy/core/src/main/scala/io/cloudstate/proxy/eventing/GooglePubsubEventing.scala

```diff
@@ -193,162 +193,291 @@ class GCPubsubEventingSupport(config: Config, materializer: Materializer) extend
 
   // Create the gRPC clients used to communicate with Google Pubsub
   final val (subscriberClient, publisherClient) = {
-    implicit val m = materializer
-    implicit val s = materializer.system
-    implicit val d = s.dispatcher
-
     val clientSettings = settings.createClientSettings()
 
     // We're reusing the same clients for all communication
     val subscriberClient = ScalaSubscriberClient(clientSettings)
     val publisherClient = ScalaPublisherClient(clientSettings)
 
     // Make sure that we don't leak connections
-    s.registerOnTermination(subscriberClient.close())
-    s.registerOnTermination(publisherClient.close())
+    system.registerOnTermination(subscriberClient.close())
+    system.registerOnTermination(publisherClient.close())
 
     (subscriberClient, publisherClient)
   }
 
   private[this] val batchResults =
     if (downstreamBatchDeadline > 0.seconds || downstreamBatchSize > 1) {
-      Flow[ProtobufAny]
-        .map(any => PubsubMessage(data = any.toByteString))
+      Flow[DestinationEvent]
+        .map(transformDestinationEvent)
         .groupedWithin(downstreamBatchSize, downstreamBatchDeadline)
-    } else Flow[ProtobufAny].map(any => PubsubMessage(data = any.toByteString) :: Nil)
+    } else Flow[DestinationEvent].map(event => transformDestinationEvent(event) :: Nil)
 
-  private[this] final def createManualSource(
+  private def sourceToSourceToFlow[In, Out, MOut](f: Source[In, NotUsed] => Source[Out, MOut]): Flow[In, Out, NotUsed] =
+    Flow[In].prefixAndTail(0).flatMapConcat { case (Nil, in) => f(in) }
```

Comment:
Seems awkward for akka streams if the best way to connect this is to use empty `prefixAndTail` and substreams. Don't know of a more direct way.

Reply:
Related issues:

https://github.com/akka/akka/issues/26592
https://github.com/akka/akka-grpc/issues/550

Reply:
:+1: good to see that there are issues already. Both make sense to have, the utility method and the flow API in akka-grpc. I guess the alternative to the substreams approach is to have a reactive streams publisher/subscriber boundary to connect, although that probably means using the pre-materialize approach as well.